core/src/test/java/org/apache/iceberg/io/TestDataWriter.java

```diff
@@ -0,0 +1,97 @@
+ *  * distributed with this work for additional information
+ *  * regarding copyright ownership.  The ASF licenses this file
+ *  * to you under the Apache License, Version 2.0 (the
+ *  * "License"); you may not use this file except in compliance
+ *  * with the License.  You may obtain a copy of the License at
+ *  *
+ *  *   http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing,
+ *  * software distributed under the License is distributed on an
+ *  * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  * KIND, either express or implied.  See the License for the
+ *  * specific language governing permissions and limitations
+ *  * under the License.
+ *
+ */
+
+package org.apache.iceberg.io;
+
+import org.apache.iceberg.*;
+import org.apache.iceberg.encryption.EncryptionKeyMetadata;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+
+@RunWith(MockitoJUnitRunner.class)
+public class TestDataWriter {
+
+    DataWriter<String> dataWriter;
+    @Mock
+    FileAppender<String> appender;
+    @Mock
+    StructLike partition;
+    @Mock
+    EncryptionKeyMetadata keyMetadata;
+    @Rule
+    public TemporaryFolder temp = new TemporaryFolder();
+    private File file;
+
+    @Before
+    public void setUp() throws IOException {
+        file = temp.newFile();
+        file.deleteOnExit();
+        dataWriter = new DataWriter<>(appender, FileFormat.PARQUET, file.getPath(), PartitionSpec.unpartitioned(), partition, keyMetadata);
+    }
+
+    @Test
+    public void write() {
+        String row = "first_row";
+        dataWriter.write(row);
+        Mockito.verify(appender, Mockito.times(1)).add(row);
+    }
+
+    @Test
+    public void close() throws IOException {
+        Mockito.when(appender.length()).thenReturn(1L);
+        Mockito.when(appender.metrics()).thenReturn(Mockito.mock(Metrics.class));
+        Mockito.when(appender.splitOffsets()).thenReturn(Arrays.asList(1L));
+        dataWriter.close();
+        DataFile dataFile = dataWriter.toDataFile();
+        Assert.assertEquals(file.getPath(), dataFile.path());
+    }
+
+    @Test(expected=IllegalStateException.class)
+    public void closeFailureShouldNotReturnDataFile() throws IOException {
+        Mockito.doThrow(new IllegalStateException("mock close failure of appender"))
+                .when(appender).close();
+        try {
+            dataWriter.close();
+        } catch (Exception e) {
+            // ignore mocked failure on first call to close
+        }
+        dataWriter.close();
+        dataWriter.toDataFile();
```

Comment:
L89 : will this line execute? line 88 will always throw exception, right?

Reply:
yes, L89 will execute and we are verifying that here with expectation that call on L89 will throw an exception as expected.